engines/e_pkcs11.c

```diff
@@ -0,0 +1,490 @@
+                              &signatureLen);
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_Sign failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_RSA_ENC, PKCS11_R_SIGN_FAILED);
+        goto err;
+    }
+
+    pkcs11_logout(pkcs11st.session);
+    pkcs11_end_session(pkcs11st.session);
+    pkcs11_finalize();
+    return 1;
+
+ err:
+    return 0;
+}
+
+RSA_METHOD *pkcs11_rsa()
+{
+    static RSA_METHOD *pkcs11_rsa = NULL;
+    pkcs11_rsa = RSA_meth_new("PKCS#11 RSA method", 0);
+    RSA_meth_set_priv_enc(pkcs11_rsa, pkcs11_rsa_enc);
+    return pkcs11_rsa;
+}
+
+/**
+ * Load the PKCS#11 functions into global function list.
+ * @param library_path
+ * @return
+ */
+CK_RV pkcs11_load_functions(char *library_path)
+{
+    CK_RV rv;
+    static DSO *pkcs11_dso = NULL;
+
+    CK_RV(*pFunc)();
+    pkcs11_dso = DSO_load(NULL, library_path, NULL, 0);
+
+    if (pkcs11_dso == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_LOAD_FUNCTIONS,
+                  PKCS11_R_LIBRARY_PATH_NOT_FOUND);
+        return CKR_GENERAL_ERROR;
+    }
+
+    pFunc = (CK_RV (*)()) DSO_bind_func(pkcs11_dso, "C_GetFunctionList");
+
+    if (pFunc == NULL) {
+        PKCS11_trace("C_GetFunctionList() not found in module %s\n",
+                     library_path);
+        PKCS11err(PKCS11_F_PKCS11_LOAD_FUNCTIONS,
+                  PKCS11_R_GETFUNCTIONLIST_NOT_FOUND);
+        return CKR_FUNCTION_NOT_SUPPORTED;
+    }
+
+    rv = pFunc(&pkcs11_funcs);
+
+    return rv;
+}
+
+/**
+ * Initialize the PKCS#11 library.
+ * This loads the function list and initializes PKCS#11.
+ * @param library_path
+ * @return
+ */
+CK_RV pkcs11_initialize(char *library_path)
+{
+    CK_RV rv;
+
+    if (!library_path) {
+        return CKR_ARGUMENTS_BAD;
+    }
+
+    rv = pkcs11_load_functions(library_path);
+    if (rv != CKR_OK) {
+        PKCS11_trace("Getting PKCS11 function list failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_INITIALIZE,
+                  PKCS11_R_GETTING_FUNCTION_LIST_FAILED);
+        return rv;
+    }
+
+    CK_C_INITIALIZE_ARGS args;
+    memset(&args, 0, sizeof(args));
+    args.flags = CKF_OS_LOCKING_OK;
+    rv = pkcs11_funcs->C_Initialize(&args);
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_Initialize failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_INITIALIZE, PKCS11_R_INITIALIZE_FAILED);
+        return rv;
+    }
+
+    return CKR_OK;
+}
+
+static void pkcs11_finalize()
+{
+    pkcs11_funcs->C_Finalize(NULL);
+}
+
+static CK_SLOT_ID pkcs11_get_slot()
+{
+    CK_RV rv;
+    CK_SLOT_ID slotId;
+    CK_ULONG slotCount = 10;
+    CK_SLOT_ID *slotIds = OPENSSL_malloc(sizeof(CK_SLOT_ID) * slotCount);
+    rv = pkcs11_funcs->C_GetSlotList(CK_TRUE, slotIds, &slotCount);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_GetSlotList failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_SLOT, PKCS11_R_GET_SLOTLIST_FAILED);
+        goto err;
+    }
+
+    if (slotCount < 1) {
```

Comment:
This really means slotCount == 0 (slotCount is CK_ULONG) so make it say that directly.